### PR TITLE
Bump python/docker, switch to Github Actions

### PR DESCRIPTION
**Pull request by Mojken**

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -4,48 +4,40 @@
...
 [testenv]
 passenv = COVERAGE_FILE
 whitelist_externals = make
 commands =
     make test
 install_command =
-    pip install --no-binary --pre {opts} {packages}
+    pip install --no-binary :all: --pre {opts} {packages}
 deps = six
        Pillow<=3.4.2
```
Comment (Swamii) [inline tox.ini:21]: Another Pillow culprit here

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        uses: 5monkeys/cobertura-action@master
+        with:
+          path: coverage.xml
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          minimum_coverage: 85
+          show_line: true
+          show_branch: true
+          show_missing: true
+          link_missing_lines: true
+          link_missing_lines_source_dir: "src"
```
Comment (Swamii) [inline .github/workflows/main.yml:50]: We don't have a `src`-dir. Not sure what it should be

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Generate coverage file
+        if: matrix.python-version == '3.9'
+        run: "TOXENV=coverage python -m tox"
+      - name: Report coverage
+        if: matrix.python-version == '3.9' && github.event_name == 'pull_request' && (github.event.action == 'opened' || github.event.action == 'reopened' || github.event.action == 'synchronize')
+        uses: 5monkeys/cobertura-action@master
+        with:
+          path: coverage.xml
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          minimum_coverage: 85
```
Comment (Swamii) [inline .github/workflows/main.yml:45]: I think we should go for 93 (last coverage I found)

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        run: python -m pip install --upgrade pip setuptools virtualenv
+      - name: Install dependencies
+        run: python -m pip install --upgrade tox
+      - name: Run tox targets for ${{ matrix.python-version }}
+        run: |
+          ENV_PREFIX=$(tr -C -d "0-9" <<< "${{ matrix.python-version }}")
+          TOXENV=$(tox --listenvs | grep "^py$ENV_PREFIX" | tr '\n' ',') python -m tox
+      - name: Generate coverage file
+        if: matrix.python-version == '3.9'
+        run: "TOXENV=coverage python -m tox"
```
Comment (Swamii) [inline .github/workflows/main.yml:38]: We need to run `coverage xml` (maybe in makefile) I think (or however django-bananas does it)

Comment (Mojken) [reply .github/workflows/main.yml:38]: Main.yml is taken directly from django-bananas. It's true that these parts likely need modification!

File: .github/workflows/main.yml (hunk 2 above)
Comment (Mojken) [reply .github/workflows/main.yml]: Or perhaps it should be `djedi` or `djedi/src`

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        uses: 5monkeys/cobertura-action@master
+        with:
+          path: coverage.xml
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          minimum_coverage: 93
+          show_line: true
+          show_branch: true
+          show_missing: true
+          link_missing_lines: true
+          link_missing_lines_source_dir: "."
```
Comment (Swamii) [inline .github/workflows/main.yml:50]:
```suggestion
          link_missing_lines_source_dir: "djedi"
```

File: README.md [hunk 6]
```diff
@@ -2,7 +2,6 @@
 
 Django content management as it should be.
 
-[![Build Status](https://travis-ci.org/5monkeys/djedi-cms.svg?branch=master)](https://travis-ci.org/5monkeys/djedi-cms)
```
Comment (Swamii) [inline README.md:5]: Think we should use something like this:
`[Build Status](https://github.com/5monkeys/djedi-cms/actions)`

Same links in appropriate places in long_description in `setup.py` as well

**Review by Swamii — COMMENTED** [review]
Almost 😄

**Review by Swamii — APPROVED** [review]
👍

File: setup.py [hunk 7]
```diff
@@ -1,70 +1,70 @@
...
+        "Environment :: Web Environment",
+        "Natural Language :: English",
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: BSD License",
+        "Operating System :: OS Independent",
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
```
Comment (lundberg) [inline setup.py:60]: 3.6?

File: tox.ini [hunk 8]
```diff
@@ -4,48 +4,43 @@
...
-envlist = py27-django{ 15, 16, 17, 18, 19, 110, 111             },
-          py35-django{             18, 19, 110, 111, 20, 21, 22 },
-          py36-django{                          111, 20, 21, 22 },
-          py37-django{                          111, 20, 21, 22 },
-          py38-django{                          111, 20, 21, 22 }
+envlist = py36-django{ 22, 30, 31, 32 },
+          py37-django{ 22, 30, 31, 32 },
+          py38-django{ 22, 30, 31, 32 },
+          py39-django{ 22, 30, 31, 32 }
+
```
Comment (lundberg) [inline tox.ini:11]: Have we tried adding django 4?

File: setup.py (hunk 7 above)
Comment (Mojken) [reply setup.py]: I removed it and then forgot to add it back in when we added support again!

File: tox.ini (hunk 8 above)
Comment (Mojken) [reply tox.ini:11]: I have now, does not work out-of-the-box, but might be worth creating a new PR for.

File: djedi/compat.py [hunk 9]
```diff
@@ -1,111 +1,89 @@
...
-    return list(urls)
+
+def generic_tag_compiler(
+    parser, token, params, varargs, varkw, defaults, name, takes_context, node_class
+):
+    """
+    Returns a template.Node subclass.
+
+    This got inlined into django.template.library since Django since 1.9, this here
+    is a copypasta replacement:
```
Comment (lundberg) [inline djedi/compat.py:41]: Seems like we can use django's now that we dont need compat

File: djedi/compat.py [hunk 10]
```diff
@@ -1,111 +1,89 @@
...
-    from django.conf.urls import include, url
+def patterns(*urls):
+    return list(urls)
 
 
-if django.VERSION < (2, 0):
-    from django.core.urlresolvers import reverse, NoReverseMatch
-else:
-    from django.urls import reverse, NoReverseMatch
+def parse_bits(parser, bits, params, varargs, varkw, defaults, takes_context, name):
```
Comment (lundberg) [inline djedi/compat.py:17]: We should be able to use the `library.parse_bits` directly in-place.

File: djedi/compat.py [hunk 11]
```diff
@@ -1,111 +1,89 @@
...
+
+
+class TemplateResponse(BaseTemplateResponse):
+    def __init__(self, *args, **kwargs):
+        kwargs["using"] = "django"
+        super(TemplateResponse, self).__init__(*args, **kwargs)
+
+
+def get_cache(name):
+    return caches[name]
```
Comment (lundberg) [inline djedi/compat.py:62]: Let's use `caches[name]` in place.

File: djedi/tests/test_templatetags.py [hunk 12]
```diff
@@ -1,104 +1,117 @@
...
+        cio.set("i18n://sv-se@page/title.txt", u"Swedish Djedi")
         html = self.render(u"{% node 'page/title' edit=False %}")
-        assert html == u'Swedish Djedi'
+        assert html == u"Swedish Djedi"
 
-        with self.settings(DJEDI={'URI_DEFAULT_SCHEME': 'l10n'}):
+        with self.settings(DJEDI={"URI_DEFAULT_SCHEME": "l10n"}):
             html = self.render(u"{% node 'page/title' edit=False %}")
-            assert html == u''
+            assert html == u""
```
Comment (lundberg) [inline djedi/tests/test_templatetags.py:54]: We should change all `u""` to `""` now that we've dropped py2

Comment (Swamii) [reply djedi/tests/test_templatetags.py:54]: We could leave it for a PR where we add pre-commit (then it's done automatically with pyupgrade, if that fits in this project)

**Comment by lundberg** [comment]
I think we should try to drop the `djedi.compat` module completely, really  no compat left, just plain django and some *"utils"* 😉

File: djedi/compat.py [hunk 13]
```diff
@@ -1,111 +1,68 @@
...
-           'url',
-           'reverse',
-           'NoReverseMatch',
-           'generic_tag_compiler',
-           'parse_bits',
-           'get_cache']
+from django.conf.urls import include, url
+from django.urls import reverse, NoReverseMatch
+from django.template.library import parse_bits
+
```
Comment (lundberg) [inline djedi/compat.py:11]: I think we should try to drop this `djedi.compat` module completely, really  no compat left, just plain django and some *"utils"* 😉

File: djedi/compat.py (hunk 9 above)
Comment (lundberg) [reply djedi/compat.py]: nvm 😂

**Review by lundberg — APPROVED** [review]
Looking good!

Side note...it would've been great if no formatting was done in this PR, made it a lot harder to review and ensure no logic/tests were touched 😅 .

**Comment by Mojken** [comment]
I've gone ahead and merged the pre-commit PR into this (hence the complete rewrite of the history), so the files changed-count will go down drastically once that PR is merged in!

**Comment by Mojken** [comment]
Closed in favor of #95

**Pull request closed without merging**